src/shell/shell_misc.cpp

```diff
@@ -410,49 +409,38 @@ bool DOS_Shell::Execute(char * name,char * args) {
 		}
 		return true;
 	}
+
 	/* Check for a full name */
-	p_fullname = Which(name);
-	if (!p_fullname) return false;
+	const char *p_fullname = Which(name);
+	if (!p_fullname)
+		return false;
 	safe_strcpy(fullname, p_fullname);
-	const char* extension = strrchr(fullname,'.');
-	
-	/*always disallow files without extension from being executed. */
-	/*only internal commands can be run this way and they never get in this handler */
-	if(extension == 0)
-	{
-		//Check if the result will fit in the parameters. Else abort
-		if(strlen(fullname) >( DOS_PATHLENGTH - 1) ) return false;
-		char temp_name[DOS_PATHLENGTH+4],* temp_fullname;
-		//try to add .com, .exe and .bat extensions to filename
-
-		safe_strcpy(temp_name, fullname);
-		safe_strcat(temp_name, ".COM");
-		temp_fullname=Which(temp_name);
-		if (temp_fullname) { extension=".com";strcpy(fullname,temp_fullname); }
-
-		else 
-		{
-			safe_strcpy(temp_name, fullname);
-			safe_strcat(temp_name, ".EXE");
-			temp_fullname=Which(temp_name);
-		 	if (temp_fullname) { extension=".exe";strcpy(fullname,temp_fullname);}
 
-			else 
-			{
-				safe_strcpy(temp_name, fullname);
-				safe_strcat(temp_name, ".BAT");
-				temp_fullname=Which(temp_name);
-		 		if (temp_fullname) { extension=".bat";strcpy(fullname,temp_fullname);}
-
-				else  
-				{
-		 			return false;
-				}
-			
-			}	
+	// Always disallow files without extension from being executed.
+	// Only internal commands can be run this way and they never get in
+	// this handler.
+	const char *extension = strrchr(fullname, '.');
+	if (extension == nullptr) {
+		// Check if the result will fit in the parameters.
+		if (strlen(fullname) > (DOS_PATHLENGTH - 1))
+			return false;
```

Comment:
I wonder if we can improve this further.  
If `fullname` is already too long to accommodate an additional "dot-three-letter-extension", then we want to skip it.  So perhaps > (DOS_PATHLENGTH - 5)? (or -4)?.. I'm not sure of the trailing size details.

Reply:
I think this check existed only to make sure that appending 4 letters will not overflow the buffer (which is still a valid concern - except in our case instead of overflow it would lead to the potential invocation of non-executable file).

Overall I would keep it as it is because I don't fully understand why DOS_PATHLENGTH in dosbox is 80; depending on source I found various claims of max path being between 66 and 260…

One comment I found referred to DOS CDS (Current Directory Structure) having a buffer of size 67 (0-delimited, so storing max path 66), but I can't find any real reference to it (inside dosbox nor in other sources). It would actually make sense - 66-long dir path + slash + (8+3) filename + 0-delimiter == 80 characters, but I don't have any documentation to confirm this ATM.